src/lxc/lxc_attach.c

```diff
@@ -273,7 +273,12 @@ static int get_pty_on_host(struct lxc_container *c, struct wrapargs *wrap, int *
 	if (lxc_console_create(conf) < 0)
 		return -1;
 	ts = conf->console.tty_state;
-	ts->stdoutfd = STDOUT_FILENO;
+	/* If stdout is not a tty but stderr is, we write to stderr. */
+	if (!isatty(STDOUT_FILENO) && isatty(STDERR_FILENO))
+		ts->stdoutfd = STDERR_FILENO;
+	else
+		ts->stdoutfd = STDOUT_FILENO;
+
```

Comment:
This is indeed confusing. So maybe a longer explanation is in order. As Serge correctly pointed out we have to allocate a new pty for `lxc-attach` when any of the standard file descriptors refers to a `pty`. This is to avoid providing an attack surface for input faking via a `TIOCSTI` `ioctl()` call.

The idea is to create a new pty master/slave pair on the host via `openpty()`. We then pass the file descriptors for the master/slave pair to our `lxc_attach()` function. `lxc_attach()` attaches to the (specified) namespaces of a running container and runs a command in it. Before, this was done via either one of two callback functions `lxc_attach_run_shell()` or `lxc_attach_run_command()`, depending on whether the user wants an interactive shell or simply run a command he specifies. In the new implementation we provide an additional wrapper around these two functions. This wrapper receives the master/slave pair file descriptors we created on the host and basically does what `login_tty()` would do. We only skip the `fork()` step since `lxc_attach()` will fork anyway. To go into a little more detail: Right after the fork we close the master file descriptor of the pty since we do not need it anymore. Afterwards we do `setsid()` to become session leader and `ioctl(TIOCSTTY)` on the slave file descriptor of the pty to become the controlling terminal of the session. Finally, and this is the crucial step here, we `dup2()` any standard file descriptors (stdin, stdout, and stderr) that refer to a pty on the host. It is important to note that this is an iff: If-and-only-if the standard file descriptor in question refers to a pty on the host, we `dup2()` it. If not we leave it untouched. This means redirection should stay intact. In case all file descriptors refer to a pty stdout and stderr are mixed and there is no way to separate them (This however, should not be a problem since that simply means that no redirection occurs in most cases. If I'm not mistaken this is similar to an ssh session that allocates a pty.) However, if an fd like stderr does not refer to a pty it is not `dup2()`ed. Thus, in an example like:

```
lxc-attach -n a -- sh -c "hostname >&2" > /dev/null
```

the following happens:
- We create a new pty on the host and receive a master/slave fd pair.
- We pass the master/slave fd pair to our wrapper and send it to `lxc_attach()` to execute in the containers namespaces.
- The wrapper checks which standard file descriptors refer to a pty and `dup2()`s those. For our example this means: The wrapper will find that stdin and stderr refer to a terminal an `dup2()`s them. Since stdout, however, is redirected and does not refer to a pty it will not be `dup2()` and refer to `/dev/null`. Within the container the synthetic reproducer created via `/bin/sh -c "hostname >&2"` generates output on stderr. We also want to receive that ouput on stderr on the host. The little trick in the code we pull is to check if stdout refers to a pty and if so we set `ts->stdoutfd` to stdout. If however, we detect that stdout does not refer to a pty we test whether stderr still does and if so set `ts->stdoutfd` to stderr. This will allow us to receive output on the correct standard file descriptor as the command in the container sent output to.

Now, with regards to `ts`. `ts` is a structure we use to keep all information regarding pty setup to `epoll()` on in a mainloop. This mainloop has two i/o handlers: `lxc_console_cb_tty_stdin()` and `lxc_console_cb_tty_master()`. `*_tty_stdin()` will wait for input on stdin (refered to via `ts->stdinfd` and write to `ts->masterfd`. `*_cb_tty_master()` will then receive the input on `ts->masterfd` and write to `ts->stdoutfd` which we have setup appropriately before (to stdout or stderr). In case no redirection occurs at all, `ts->stdoutfd` can be safely set to stdout since the stdout and stderr streams are mixed anyway. I hope this helps to clarify a little.